algorithms/linfa-logistic/src/lib.rs

```diff
@@ -159,45 +159,38 @@ impl<F: Float> LogisticRegression<F> {
 
     /// Ensure that `x` and `y` have the right shape and that all data and
     /// configuration parameters are finite.
-    fn validate_data<A, B>(
-        &self,
-        x: &ArrayBase<A, Ix2>,
-        y: &ArrayBase<B, Ix1>,
-    ) -> Result<(), String>
+    fn validate_data<A, B>(&self, x: &ArrayBase<A, Ix2>, y: &ArrayBase<B, Ix1>) -> Result<()>
     where
         A: Data<Elem = F>,
         B: Data<Elem = F>,
     {
         if x.shape()[0] != y.len() {
-            return Err(
-                "Incompatible shapes of data, expected `x` and `y` to have same number of rows"
-                    .to_string(),
-            );
+            return Err(Error::MismatchedShapes);
```

Comment:
you can also pass the number of rows and print that in your error, like "Expected same shape, got <..> ≠ <..>"

Reply:
👍🏻